src/hotspot/share/opto/divnode.cpp

```diff
@@ -144,118 +142,115 @@ static Node *transform_int_divide( PhaseGVN *phase, Node *dividend, jint divisor
-// minor type name and parameter changes.  Adjusted to 64 bit word width.
-static bool magic_long_divide_constants(jlong d, jlong &M, jint &s) {
-  int64_t p;
-  uint64_t ad, anc, delta, q1, r1, q2, r2, t;
-  const uint64_t two63 = UCONST64(0x8000000000000000);     // 2**63.
-
-  ad = ABS(d);
-  if (d == 0 || d == 1) return false;
-  t = two63 + ((uint64_t)d >> 63);
-  anc = t - 1 - t%ad;     // Absolute value of nc.
-  p = 63;                 // Init. p.
-  q1 = two63/anc;         // Init. q1 = 2**p/|nc|.
-  r1 = two63 - q1*anc;    // Init. r1 = rem(2**p, |nc|).
-  q2 = two63/ad;          // Init. q2 = 2**p/|d|.
-  r2 = two63 - q2*ad;     // Init. r2 = rem(2**p, |d|).
-  do {
-    p = p + 1;
-    q1 = 2*q1;            // Update q1 = 2**p/|nc|.
-    r1 = 2*r1;            // Update r1 = rem(2**p, |nc|).
-    if (r1 >= anc) {      // (Must be an unsigned
-      q1 = q1 + 1;        // comparison here).
-      r1 = r1 - anc;
-    }
-    q2 = 2*q2;            // Update q2 = 2**p/|d|.
-    r2 = 2*r2;            // Update r2 = rem(2**p, |d|).
-    if (r2 >= ad) {       // (Must be an unsigned
-      q2 = q2 + 1;        // comparison here).
-      r2 = r2 - ad;
+  juint magic_const;
+  bool magic_const_ovf;
+  juint shift_const;
+  magic_divide_constants<juint>(divisor, 0, max_pos, 0, magic_const, magic_const_ovf, shift_const);
+
+  if (!magic_const_ovf && juint_mul_no_ovf(max_pos, magic_const)) {
+    // If x * c can fit into a u32, use int multiplication
+    Node* mul = phase->transform(new MulINode(dividend, phase->intcon(magic_const)));
+    return new URShiftINode(mul, phase->intcon(shift_const));
+  }
+
+  julong magic_const_long = julong(magic_const) + (magic_const_ovf ? julong(max_juint) + 1 : 0);
+  // Unsigned extension of dividend
+  Node* dividend_long = phase->transform(new ConvI2LNode(dividend));
+  dividend_long = phase->transform(new AndLNode(dividend_long, phase->longcon(max_juint)));
+
+  if (julong_mul_no_ovf(julong(max_pos), magic_const_long)) {
+    // If x * c can fit into a u64, use long multiplication
+
+    // Java shifts are modular so we need this special case
+    if (shift_const == N * 2) {
+      return new ConINode(TypeInt::ZERO);
     }
-    delta = ad - r2;
-  } while (q1 < delta || (q1 == delta && r1 == 0));
 
-  M = q2 + 1;
-  if (d < 0) M = -M;      // Magic number and
-  s = p - 64;             // shift amount to return.
+    // q = (x * c) >> s
+    Node* mul = phase->transform(new MulLNode(dividend_long, phase->longcon(magic_const_long)));
+    Node* q = phase->transform(new URShiftLNode(mul, phase->intcon(shift_const)));
+    return new ConvL2INode(q);
+  }
+
+  // Original plan fails, rounding up of 1/divisor does not work, change
+  // to rounding down, now it is guaranteed to be correct, according to
+  // N-Bit Unsigned Division Via N-Bit Multiply-Add by Arch D. Robison
+  magic_divide_constants_round_down(divisor, magic_const, shift_const);
```

Comment:
I think there's no need for `magic_divide_constants_round_down()`.

Firstly, we recover the previous value of `magic_const` and `shift_const`, just before the overflow:
```
  magic_const = magic_const + 1 >> 1 | 0x8000_0000;
  shift_const -= 1;
```
Then we decrement `magic_const` by one:
```
  magic_const -= 1;
```
That's it.
If desired, we can additionally reduce `magic_const` to an odd value by right-shifting it by the number of trailing zero bits, and updating `shift_const` accordingly. But for the usage here, I think it makes no real difference.

We can thus avoid the division in `magic_divide_constants_round_down()`, and can get rid of the method altogether, as it seems to be used only here.
This might even open similar code for the `julong` case.

Reply:
@rgiulietti I think that is a brilliant observation. However, I think that doing so would not reduce the code complexity and it also puts another burden of proof on us for the mathematical lemma.

Regarding the `julong` case. rounding down is interesting, as we can emit code like this:

    addq dividend, 1
    je overflow
    ... (the remaining calculation)

And if `dividend` is known to be different from `max_julong` then the `je` can be removed.

As a result, I think it would be possible to address your idea in a next RFE. As I think the patch is dense enough in mathematical proofs.

Thanks a lot.